sw/device/silicon_creator/lib/cert/cert.c

```diff
@@ -5,14 +5,55 @@
 #include "sw/device/silicon_creator/lib/cert/cert.h"
 
 #include "sw/device/lib/base/hardened.h"
+#include "sw/device/silicon_creator/lib/base/util.h"
 #include "sw/device/silicon_creator/lib/cert/asn1.h"
 #include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/drivers/hmac.h"
+#include "sw/device/silicon_creator/lib/drivers/keymgr.h"
 #include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/otbn_boot_services.h"
+#include "sw/device/silicon_creator/lib/sigverify/ecdsa_p256_key.h"
 
 #include "flash_ctrl_regs.h"  // Generated.
 
 static uint8_t actual_serial_number[kCertX509Asn1SerialNumberSizeInBytes] = {0};
 
+/**
+ * Helper function to convert an attestation public key from little to big
+ * endian in place.
+ */
+static void curr_pubkey_le_to_be_convert(ecdsa_p256_public_key_t *pubkey) {
+  util_le_be_buf_format((unsigned char *)pubkey->x,
+                        kEcdsaP256PublicKeyCoordBytes);
+  util_le_be_buf_format((unsigned char *)pubkey->y,
+                        kEcdsaP256PublicKeyCoordBytes);
+}
+
+rom_error_t cert_ecc_p256_keygen(sc_keymgr_ecc_key_t key,
+                                 hmac_digest_t *pubkey_id,
+                                 ecdsa_p256_public_key_t *pubkey) {
+  HARDENED_RETURN_IF_ERROR(sc_keymgr_state_check(key.required_keymgr_state));
+
+  // Generate / sideload key material into OTBN, and generate the ECC keypair.
+  HARDENED_RETURN_IF_ERROR(otbn_boot_attestation_keygen(
+      key.keygen_seed_idx, key.type, *key.keymgr_diversifier, pubkey));
+
+  // Keys are represented in certificates in big endian format, but the key is
+  // output from OTBN in little endian format, so we convert the key to
+  // big endian format.
+  curr_pubkey_le_to_be_convert(pubkey);
+
+  // Generate the key ID.
+  //
+  // Note: the certificate generation functions expect the digest to be in big
+  // endian form, but the HMAC driver returns the digest in little endian, so we
+  // re-format it.
+  hmac_sha256(pubkey, kEcdsaP256PublicKeyCoordBytes * 2, pubkey_id);
```

Comment:
Considering that `pubkey` is the struct containing the X and Y values, and nothing else, you could use sizeof here:
```
hmac_sha256(pubkey, sizeof(*pubkey), pubkey_id);
util_le_be_buf_format(pubkey_id, sizeof(*pubkey_id));
```

Reply:
fixed.